File: generated_code/python/solution_1381_3.py
from typing import List
from collections import Counter

class Solution:
    def maxScoreWords(self, words: List[str], letters: List[str], score: List[int]) -> int:
        # Count available letter frequencies
        letter_counts = Counter(letters)
        
        # Precompute the word counts and their potential individual scores
        word_data = []
        for word in words:
            word_count = Counter(word)
            
            # Check if the word is even possible to form with total available letters
            possible = True
            word_score = 0
            for char, count in word_count.items():
                if count > letter_counts[char]:
                    possible = False
                    break
                word_score += score[ord(char) - ord('a')] * count
            
            if possible:
                word_data.append((word_count, word_score))
                
        num_words = len(word_data)
        self.max_score = 0
        
        # Backtracking function to explore subsets of valid words
        def backtrack(index: int, current_score: int, available_letters: Counter):
            # Update global maximum score found so far
            if current_score > self.max_score:
                self.max_score = current_score
                
            # Base case: explored all valid words
            if index == num_words:
                return
            
            # Option 1: Skip the current word
            backtrack(index + 1, current_score, available_letters)
            
            # Option 2: Try to include the current word
            word_count, word_score = word_data[index]
            can_form = True
            
            # Verify if the word can be formed with the CURRENT available letters
            for char, count in word_count.items():
                if available_letters[char] < count:
                    can_form = False
                    break
                    
            if can_form:
                # Deduct letters used by the current word
                for char, count in word_count.items():
                    available_letters[char] -= count
                    
                # Recurse with the word included
                backtrack(index + 1, current_score + word_score, available_letters)
                
                # Backtrack: restore the deducted letters
                for char, count in word_count.items():
                    available_letters[char] += count

        backtrack(0, 0, letter_counts)
        return self.max_score

File: generated_code/cpp/solution_1381_3.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxScoreWords(vector<string>& words, vector<char>& letters, vector<int>& score) {
        // Frequency array for the available letters
        vector<int> count(26, 0);
        for (char c : letters) {
            count[c - 'a']++;
        }
        
        // Since words.length <= 14, backtracking/DFS exploring all subsets is optimal.
        // Time Complexity: O(2^N * L), where N is number of words and L is max length of a word.
        // Space Complexity: O(N) for the recursion stack.
        return backtrack(0, words, count, score);
    }

private:
    int backtrack(int index, const vector<string>& words, vector<int>& count, const vector<int>& score) {
        if (index == words.size()) {
            return 0;
        }

        // Option 1: Skip the current word
        int max_score = backtrack(index + 1, words, count, score);

        // Option 2: Try to include the current word
        const string& word = words[index];
        bool can_form = true;
        int current_word_score = 0;
        
        // Form the word and check if we have enough letters
        for (char c : word) {
            int char_idx = c - 'a';
            count[char_idx]--;
            current_word_score += score[char_idx];
            if (count[char_idx] < 0) {
                can_form = false;
            }
        }

        // If the word can be successfully formed, explore further
        if (can_form) {
            max_score = max(max_score, current_word_score + backtrack(index + 1, words, count, score));
        }

        // Backtrack: restore the letter frequencies
        for (char c : word) {
            count[c - 'a']++;
        }

        return max_score;
    }
};